src/maths.f90

```diff
@@ -2153,11 +2244,543 @@ FUNCTION COTH_DP(A)
+  END SUBROUTINE s3_fs
+
+  !
+  !================================================================================================================================
+  !
+  
+  !> Evaluates a cubic spline at a specified point. First call spline_cubic_set to calculate derivatives
+  !> algorithm adapted from John Burkhardt's spline_cubic_val routine from the SPLINE package (https://orion.math.iastate.edu/burkardt/f_src/spline/spline.html)
+  SUBROUTINE spline_cubic_val (n, t, y, ypp, tval, yval, ypval, yppval, err, error, *)
+
+    !Argument variables
+    INTEGER(INTG), INTENT(IN) :: n !< size of t,y arrays to interpolate values from
+    REAL(DP), INTENT(IN) :: t(n) !< t array: known knot values
+    REAL(DP), INTENT(IN) :: y(n) !< y array: data values to interpolate at the knots
+    REAL(DP), INTENT(IN) :: ypp(n) !< 2nd derivatives of interpolating function at t values
+    REAL(DP), INTENT(IN) :: tval !< point in t at which spline is to be evaluated
+    REAL(DP), INTENT(OUT) :: yval !< spline interpolated y value at tval
+    REAL(DP), INTENT(OUT) :: ypval !< first derivative of spline interpolated y value at tval
+    REAL(DP), INTENT(OUT) :: yppval !< second derivative of spline interpolated y value at tval
+    INTEGER(INTG), INTENT(OUT) :: err !<The error code
+    TYPE(VARYING_STRING), INTENT(OUT) :: error !<The error string
+    !Local variables
+    REAL(DP) :: dt
+    REAL(DP) :: h
+    INTEGER(INTG) :: i
+    INTEGER(INTG) :: left
+    INTEGER(INTG) :: right
+    LOGICAL :: foundInterval
+
+    CALL ENTERS("spline_cubic_val",ERR,ERROR,*999)
+
+    !  Determine the interval [T(LEFT), T(RIGHT)] that contains TVAL.
+    !  Values below T(1) or above T(N) use extrapolation.
+    foundInterval = .FALSE.
+    DO i = 2, n - 1
+      IF ( tval < t(i) ) THEN
+        foundInterval=.TRUE.
+        left = i - 1
+        right = i
+        EXIT
+      ENDIF
+    ENDDO
+    IF (foundInterval .EQV. .FALSE.) THEN
+      left = n - 1
+      right = n
+    ENDIF
+
+    !  Evaluate the polynomial.
+    dt = tval - t(left)
+    h = t(right) - t(left)
+
+    yval = y(left) &
+     &   + dt * ( ( y(right) - y(left) ) / h &
+     &          - ( ypp(right) / 6.0E+00_DP + ypp(left) / 3.0E+00_DP ) * h &
+     &   + dt * ( 0.5E+00 * ypp(left) &
+     &   + dt * ( ( ypp(right) - ypp(left) ) / ( 6.0E+00_DP * h ) ) ) )
+
+    ypval = ( y(right) - y(left) ) / h &
+     &   - ( ypp(right) / 6.0E+00_DP + ypp(left) / 3.0E+00_DP ) * h &
+     &   + dt * ( ypp(left) &
+     &   + dt * ( 0.5E+00_DP * ( ypp(right) - ypp(left) ) / h ) )
+
+    yppval = ypp(left) + dt * ( ypp(right) - ypp(left) ) / h 
+
+    CALL EXITS("spline_cubic_val")
+    RETURN
+999 CALL ERRORS("spline_cubic_val",ERR,ERROR)
+    CALL EXITS("spline_cubic_val")
+    RETURN 1
+  END SUBROUTINE spline_cubic_val
 
   !
   !================================================================================================================================
   !
 
+  SUBROUTINE J_DPC (z,J0,J1,err,error,*)
```

Comment:
Hi @dladd ,

I got them from:
http://ptsl.phy.ncu.edu.tw/~shchen/Numerics/Garcia/Fortran/
and I think they are not under a open source license!